File: spec/cpp_stl/test_switch_manual_int_size_else.cpp
#include <boost/test/unit_test.hpp>

#include <switch_manual_int_size_else.h>

#include <iostream>
#include <fstream>
#include <vector>

#include "helpers.h"

BOOST_AUTO_TEST_CASE(test_switch_manual_int_size_else) {
    std::ifstream ifs("src/switch_tlv.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    switch_manual_int_size_else_t* r = new switch_manual_int_size_else_t(&ks);

    BOOST_CHECK_EQUAL(r->chunks()->size(), 4);

    BOOST_CHECK_EQUAL(r->chunks()->at(0)->code(), 0x11);
    switch_manual_int_size_else_t::chunk_t::chunk_meta_t* meta =
        static_cast<switch_manual_int_size_else_t::chunk_t::chunk_meta_t*>(r->chunks()->at(0)->body());
    BOOST_CHECK_EQUAL(meta->title(), "Stuff");
    BOOST_CHECK_EQUAL(meta->author(), "Me");

    BOOST_CHECK_EQUAL(r->chunks()->at(1)->code(), 0x22);
    switch_manual_int_size_else_t::chunk_t::chunk_dir_t* dir =
        static_cast<switch_manual_int_size_else_t::chunk_t::chunk_dir_t*>(r->chunks()->at(1)->body());
    COMPARE_ARRAY(std::string, dir->entries(), "AAAA", "BBBB", "CCCC");

    BOOST_CHECK_EQUAL(r->chunks()->at(2)->code(), 0x33);
    switch_manual_int_size_else_t::chunk_t::dummy_t* dummy1 =
        static_cast<switch_manual_int_size_else_t::chunk_t::dummy_t*>(r->chunks()->at(2)->body());
    BOOST_CHECK_EQUAL(dummy1->rest(), std::string("\x10\x20\x30\x40\x50\x60\x70\x80"));

    BOOST_CHECK_EQUAL(r->chunks()->at(3)->code(), 0xff);
    switch_manual_int_size_else_t::chunk_t::dummy_t* dummy2 =
        static_cast<switch_manual_int_size_else_t::chunk_t::dummy_t*>(r->chunks()->at(3)->body());
    BOOST_CHECK_EQUAL(dummy2->rest(), std::string(""));

    delete r;
}

File: spec/rust/tests/test_switch_manual_int_size_else.rs
#![allow(dead_code)]
use std::{fs, rc::Rc};

extern crate kaitai;
use self::kaitai::*;
mod formats;
use formats::switch_manual_int_size_else::*;

#[test]
fn test_switch_manual_int_size_else() {
    let bytes = fs::read("../../src/switch_tlv.bin").unwrap();
    let reader = BytesReader::new(&bytes);
    let res = SwitchManualIntSizeElse::read_into(&reader, None, None);
    let r : Rc<SwitchManualIntSizeElse>;

    if let Err(err) = res {
        panic!("{:?}", err);
    } else {
        r = res.unwrap();
    }
    assert_eq!(4, r.chunks().len());

    assert_eq!(17, *r.chunks()[0].code());
    if let SwitchManualIntSizeElse_Chunk_Body::SwitchManualIntSizeElse_Chunk_ChunkMeta(s) =  r.chunks()[0].body() {
        assert_eq!("Stuff", s.title());
        assert_eq!("Me", s.author());
    } else {
        panic!("expected enum SwitchManualIntSizeElse_Chunk_ChunkMeta");
    }

    assert_eq!(34, *r.chunks()[1].code());
    if let SwitchManualIntSizeElse_Chunk_Body::SwitchManualIntSizeElse_Chunk_ChunkDir(s) =  r.chunks()[1].body() {
        let strings : Vec<String> = vec!["AAAA", "BBBB", "CCCC"].iter().map(|&s| s.to_string() ).collect();
        assert_eq!(strings, *s.entries());
    } else {
        panic!("expected enum SwitchManualIntSizeElse_Chunk_ChunkDir");
    }
 
    assert_eq!(51, *r.chunks()[2].code());
    if let SwitchManualIntSizeElse_Chunk_Body::SwitchManualIntSizeElse_Chunk_Dummy(s) =  r.chunks()[2].body() {
        let raw : Vec<u8> = vec![0x10, 0x20, 0x30, 0x40, 0x50, 0x60, 0x70, 0x80];
        assert_eq!(raw, *s.rest());
    } else {
        panic!("expected enum SwitchManualIntSizeElse_Chunk_Dummy");
    }
 
    assert_eq!(255, *r.chunks()[3].code());
    if let SwitchManualIntSizeElse_Chunk_Body::SwitchManualIntSizeElse_Chunk_Dummy(s) =  r.chunks()[3].body() {
        let raw : Vec<u8> = vec![];
        assert_eq!(raw, *s.rest());
    } else {
        panic!("expected enum SwitchManualIntSizeElse_Chunk_Dummy");
    }
}
